src/build.d

```diff
@@ -907,23 +893,51 @@ struct Dependency
     string[] targets; // list of all target files
     string[] sources; // list of all source files
     string[] rebuildSources; // Optional list of files that trigger a rebuild of this dependency
+    DependencyRef[] deps; // dependencies to build before this one
     string[] command; // the dependency command
     void delegate() commandFunction; // a custom dependency command which gets called instead of command
     string name; // name of the dependency that is e.g. written to the CLI when it's executed
     string[] trackSources;
+}
 
-    /**
-    Executes the dependency
-    */
-    auto run()
+class DependencyRef
+{
+    Dependency dep;
+    alias dep this;
+    private bool executed;
+
+    this(ref Dependency dep)
     {
-        // allow one or multiple targets
-        if (target !is null)
-            targets = [target];
+        this.dep = dep;
+        if (dep.target)
+        {
+            assert(!dep.targets, "target and targets cannot both be set");
+            this.dep.targets = [dep.target];
+        }
+    }
+
+    /// Executes the dependency
+    void run()
+    {
+        synchronized (this)
```

Comment:
double checking that you want to sync on the reference's monitor and the `dep`s monitor?

Reply:
The `dep` is a struct, it doesn't have a monitor.  Just syncing on the classes monitor.

Reply:
Oooh... I thought it was the other way around.